File: Java/misc/Order.java
package misc;

/**
 * Static Parent
 * Static Child
 * Inline Parent
 * Parent Constructor
 * Inline Child
 * Child Primary Constructor
 * Child Secondary Constructor
 */

public class Order {
    public static void main(String[] args) {
        new Children(4);
    }
}

class Parent {
    {
        System.out.println("Parent Init 1");
    }

    static {
        System.out.println("Parent Static 1");
    }

    Parent() {
        System.out.println("Parent Primary Constructor");
    }

    {
        System.out.println("Parent Init 2");
    }

    static {
        System.out.println("Parent Static 2");
    }
}

class Children extends Parent {
    private int a = 2;

    {
        System.out.println("Child Init 1");
        System.out.println(a);
    }

    static {
        System.out.println("Child Static 1");
    }

    private int b = 4;

    static {
        System.out.println("Child Static 2");
    }

    Children() {
        super();
        System.out.println("Child Primary Constructor");
    }

    {
        System.out.println("Child Init 2");
        System.out.println(b);
    }

    Children(int a) {
        this();
        System.out.println("Child Secondary Constructor");
    }
}

File: kotlin/Order.kt
/**
 * Primary Constructor Called Before Secondary Constructor
 * Constructer called after init block and initializers
 * Init block and initializers called in top to down manner
 * Parent class is called before Child class
 */
open class Parent {
    private val a = println("Parent.a")

    constructor() {
        println("Parent primary constructor")
    }

    init {
        println("Parent.init")
    }

    private val b = println("Parent.b")
}

class Child : Parent {
    val a = println("Child.a")

    init {
        println("Child.init 1")
    }

    constructor() : super() {
        println("Child primary constructor")
    }

    val b = println("Child.b")

    constructor(s: Int): this() {
        println("Child secondary constructor")
    }

    init {
        println("Child.init 2")
    }
}

fun main() {
    Child(2)
}